samcli/lib/utils/subprocess_utils.py

```diff
@@ -0,0 +1,115 @@
+"""
+Utils for invoking subprocess calls
+"""
+import logging
+from time import sleep
+
+import sys
+from subprocess import Popen
+
+from typing import Callable, Dict, Any, Optional, Union, AnyStr, IO
+
+from samcli.commands.exceptions import UserException
+from samcli.lib.utils.stream_writer import StreamWriter
+
+LOG = logging.getLogger(__name__)
+
+
+class LoadingPatternError(UserException):
+    def __init__(self, ex):
+        self.ex = ex
+        message_fmt = f"Failed to execute the subprocess. {ex}"
+        super().__init__(message=message_fmt.format(ex=self.ex))
+
+
+def default_loading_pattern(stream_writer: Optional[StreamWriter] = None, loading_pattern_rate: float = 0.5) -> None:
+    """
+    A loading pattern that just prints '.' to the terminal
+
+    Parameters
+    ----------
+    stream_writer: Optional[StreamWriter]
+        The stream to which to write the pattern
+    loading_pattern_rate: int
+        How frequently to generate the pattern
+    """
+    stream_writer = stream_writer or StreamWriter(sys.stderr)
+    stream_writer.write(".")
+    stream_writer.flush()
+    sleep(loading_pattern_rate)
+
+
+def invoke_subprocess_with_loading_pattern(
+    command_args: Dict[str, Any],
+    loading_pattern: Callable[[StreamWriter], None] = default_loading_pattern,
+    stream_writer: Optional[StreamWriter] = None,
+) -> Optional[Union[str, bytes]]:
+    """
+    Wrapper for Popen to asynchronously invoke a subprocess while
+    printing a given pattern until the subprocess is complete.
+    If the log level is lower than INFO, stream the process stdout instead.
+
+    Parameters
+    ----------
+    command_args: Dict[str, Any]
+        The arguments to give to the Popen call, should contain at least one parameter "args"
+    loading_pattern: Callable[[StreamWriter], None]
+        A function generating a pattern to the given stream
+    stream_writer: Optional[StreamWriter]
+        The stream to which to write the pattern
+
+    Returns
+    -------
+    str
+        A string containing the process output
+    """
+    stream_writer = stream_writer or StreamWriter(sys.stdout)
```

Comment:
should it be `sys.stderr`, to leave the stdout for the lambda function output?

Reply:
Yup, good catch, didn't mean to commit that.